dali/util/numpy.cc

```diff
@@ -0,0 +1,226 @@
+  Skip(ptr, name);
+  Skip(ptr, "'");
+  SkipSpaces(ptr);
+  Skip(ptr, ":");
+  SkipSpaces(ptr);
+}
+
+template <typename T = int64_t>
+T ParseInteger(const char*& ptr) {
+  char *out_ptr = const_cast<char*>(ptr);  // strtol takes a non-const pointer
+  T value = static_cast<T>(strtol(ptr, &out_ptr, 10));
+  DALI_ENFORCE(out_ptr != ptr, "Parse error: expected a number.");
+  ptr = out_ptr;
+  return value;
+}
+
+std::string ParseStringValue(const char*& input, char delim_start = '\'', char delim_end = '\'') {
+  DALI_ENFORCE(*input++ == delim_start, make_string("Expected \'", delim_start, "\'"));
+  std::string out;
+  for (; *input != '\0'; input++) {
+    if (*input == '\\') {
+      switch (*++input) {
+        case '\\':
+          out += '\\';
+          break;
+        case '\'':
+          out += '\'';
+          break;
+        case '\t':
+          out += '\t';
+          break;
+        case '\n':
+          out += '\n';
+          break;
+        case '\"':
+          out += '\"';
+          break;
+        default:
+          out += '\\';
+          out += *input;
+          break;
+      }
+    } else if (*input == delim_end) {
+      break;
+    } else {
+      out += *input;
+    }
+  }
+  DALI_ENFORCE(*input++ == delim_end, make_string("Expected \'", delim_end, "\'"));
+  return out;
+}
+
+namespace detail {
+
+void ParseHeaderMetadata(HeaderData& target, const std::string &header) {
+  target.shape = {};
+  const char* hdr = header.c_str();
+  SkipSpaces(hdr);
+  Skip(hdr, "{");
+  SkipFieldName(hdr, "descr");
+  auto typestr = ParseStringValue(hdr);
+    // < means LE, | means N/A, = means native. In all those cases, we can read
+  bool little_endian = (typestr[0] == '<' || typestr[0] == '|' || typestr[0] == '=');
+  DALI_ENFORCE(little_endian, "Big Endian files are not supported.");
+  target.type_info = &TypeFromNumpyStr(typestr.substr(1));
+
+  SkipSpaces(hdr);
+  Skip(hdr, ",");
+  SkipFieldName(hdr, "fortran_order");
+  if (TrySkip(hdr, "True")) {
+    target.fortran_order = true;
+  } else if (TrySkip(hdr, "False")) {
+    target.fortran_order = false;
+  } else {
+    DALI_FAIL("Failed to parse fortran_order field.");
+  }
+  SkipSpaces(hdr);
+  Skip(hdr, ",");
+  SkipFieldName(hdr, "shape");
+  Skip(hdr, "(");
+  SkipSpaces(hdr);
+  while (*hdr != ')') {
+    // ParseInteger already skips the leading spaces (strtol does).
+    target.shape.shape.push_back(ParseInteger<int64_t>(hdr));
+    SkipSpaces(hdr);
+    DALI_ENFORCE(TrySkip(hdr, ",") || target.shape.size() > 1,
+                 "The first number in a tuple must be followed by a comma.");
+  }
+  if (target.fortran_order) {
+    // cheapest thing to do is to define the tensor in an reversed way
+    std::reverse(target.shape.begin(), target.shape.end());
+  }
+}
+
+}  // namespace detail
+
+HeaderData ParseHeader(InputStream *src) {
+  HeaderData parsed_header;
+
+  // check if the file is actually a numpy file
+  std::vector<uint8_t> token(11);
```

Comment:
```suggestion
  SmallVector<char, 128> token;
  token.resize(11);
```
and get rid of all the `reinterpret_cast<char *>`

Reply:
Fixed in https://github.com/NVIDIA/DALI/pull/4100/commits/76506b1875d8c460fd9bbfc193911a2cd71945e4